api/app/Traits/Generator/Filterable.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace App\Traits\Generator;
+
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Support\Arr;
+
+trait Filterable
+{
+    protected ?array $filters;
+
+    protected ?array $ids;
+
+    /**
+     * Set the filters to be applied
+     *
+     * @param  ?array  $filters  The filters being applied
+     * @return GeneratorIsFilterable The class instance
+     */
+    public function setFilters(?array $filters)
+    {
+        $this->filters = $filters;
+
+        return $this;
+    }
+
+    /**
+     * Set the ids to be applied
+     *
+     * @param  ?array  $ids  The ids being applied
+     * @return GeneratorIsFilterable The class instance
+     */
+    public function setIds(?array $ids)
+    {
+        $this->ids = $ids;
+
+        return $this;
+    }
+
+    /**
+     * Flattens out filters so it can be used
+     * to apply the appropriate scopes
+     *
+     * @return array The flattened filters
+     */
+    private function flattenFilters($filters): array
+    {
+        $flattened = [];
+        foreach ($filters as $k => $v) {
+            // NOTE: Equity is expected to be assoc array so do not flatten it
+            if (is_array($v) && Arr::isAssoc($v) && $k !== 'equity') {
+                $flattened = array_merge($flattened, $this->flattenFilters($v));
+            } else {
+                $flattened[$k] = $v;
+            }
+        }
+
+        return $flattened;
+    }
+
+    /**
+     * Apply filters to the query builder
+     *
+     * @param  Builder  $query  The query builder to apply filters to
+     * @param  class-string  $class  The model for the query builder
+     * @param  array  $scopeMap  A map of $filterKey => $scopeName
+     * @return Builder $query The query builder with scopes applied
+     */
+    public function applyFilters(Builder $query, string $class, ?array $scopeMap): Builder
+    {
+        if (is_null($this->filters) && is_null($this->ids)) {
+            return $query;
+        }
+
+        $filters = $this->flattenFilters($this->filters ?? []);
+        foreach ($filters as $key => $value) {
+            $scope = $scopeMap[$key] ?? $key;
+
+            if (method_exists($class, 'scope'.ucfirst($scope)) && $value) {
+                $query?->$scope($value);
+            }
+        }
+
```

Comment:
Appears to fail to apply skill filters, I think it might be due to the names

Adding logging

![image](https://github.com/user-attachments/assets/afe99b92-13d7-4afa-a378-54336961ad50)

Logs - filters

![image](https://github.com/user-attachments/assets/52c240bf-8d9d-44ab-a2b2-29bd5c36fab4)

Logs - skill filter

![image](https://github.com/user-attachments/assets/7597b9a8-727a-4500-a236-25e1ec23a44d)

Reply:
I'm pretty sure skills don't actually filter do they? It is a really weird scope that just tells you how many skills the user has that matches. I could be wrong though.

Does the document downloaded match the number of rows in the table?

Reply:
Nvm, I just tested myself and it actually does filter them

Reply:
Also, seems to work for me :sweat_smile: 

![2024-08-21_15-22](https://github.com/user-attachments/assets/fc148ca9-bef6-44af-b008-e224fbadcaba)

Reply:
It does

![image](https://github.com/user-attachments/assets/a55980b6-4c1c-4c89-9063-991968e3a2b9)

![image](https://github.com/user-attachments/assets/7efc55b1-64b5-437d-8909-ad9c496a0710)

But the CSV contains all

Reply:
Users table, I should've mentioned

Reply:
oops, okay. Thansk, I'll take a look at that :ok_hand:

Reply:
Fixed in [fix skills filter for user csv](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11300/commits/a94ffad1d8aef31bf0955a9dd45116ee2423a6cf)